[HUDI-308] Avoid Renames for tracking state transitions of all actions on dataset

bvaradar opened the pull request:
[HUDI-308] Avoid Renames for tracking state transitions of all actions on dataset

With this PR,  Hudi Timeline management no longer uses rename to mark state transitions.  As renames can be non-atomic in some cloud stores, this PR addresses this issue in a clean way.

Related Changes:

1.  Introduce new metadata layout version to Hudi table properties and use this to determine if renames should be used or not while writing. Any existing table created prior to 0.5.1 will preserve old semantics. Newer tables that are created after 0.5.1 will automatically avoid renames. Hudi Query Engine integration should be able to handle both cases. We expect the deployment to first upgrade query engines before upgrading writer

2. As the new format enforces write once semantics, there is no longer any need to write compaction and cleaner plan in both places (.hoodie and .hoodie/.aux). Code changes handles this

3. Commits/DeltaCommits also follow requested -> inflight -> completed state transitions. Rollback for "requested" state (failure during index lookup) is trivial as no side-effects happened. 

4. Commit Archiving handles the case of intermediate state files also being present

bvaradar commented:
@vinothchandar @n3nash : Ready for review.

n3nash commented on hudi-client/src/main/java/org/apache/hudi/HoodieCleanClient.java:
```diff
@@ -92,7 +92,7 @@ protected HoodieCleanMetadata clean(String startCleanTime) throws HoodieIOExcept
           && !cleanerPlan.getFilesToBeDeletedPerPartition().isEmpty()) {
         final HoodieTable<T> hoodieTable = HoodieTable.getHoodieTable(createMetaClient(true), config, jsc);
-        return runClean(hoodieTable, startCleanTime);
+        return runClean(hoodieTable, HoodieTimeline.getCleanRequestedInstant(startCleanTime), cleanerPlan);
```
Does the getCleanRequestedInstant return REQUESTED and INFLIGHT both ?

n3nash commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -731,6 +739,12 @@ public boolean rollbackToSavepoint(String savepointTime) {
    * file,
    */
   public boolean rollback(final String commitTime) throws HoodieRollbackException {
+    try {
```
we can remove this now

n3nash commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -824,14 +843,14 @@ private String startInstant() {
 
-    List<String> inflights =
-        inflightCommitTimeline.getInstants().map(HoodieInstant::getTimestamp).collect(Collectors.toList());
+    List<String> inflights = inflightAndRequestedCommitTimeline.getInstants().map(HoodieInstant::getTimestamp)
```
Does rollback need inflight and requested both ? The restore command probably needs both but rollback should only consider inflight ?

n3nash replied:
So, that's true for other types of actions but for compaction we need to also delete the extra log file created (I'm guessing this part of the code works as expected already and isn't changed as part of this diff, correct ? )

bvaradar replied:
Yes, it should work fine. For restoring (reverting completed) compactions, we would still delete log files as part of first rolling-back the inflight state of the compaction before we try to rollback requested state.

n3nash commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -896,17 +919,22 @@ private void finishRestore(final Timer.Context context, Map<String, List<HoodieR
+          .collect(Collectors.toList());
+      if (!rollbackInstants.isEmpty()) {
+        Preconditions.checkArgument(rollbackInstants.size() == 1);
+        List<HoodieRollbackStat> stats = doRollbackAndGetStats(rollbackInstants.get(0));
```
Instead of checking and performing preconditions just do table.getActiveTimeline().getCommitsTimeline().getInstants()
          .filter(instant -> HoodieActiveTimeline.EQUAL.test(instant.getTimestamp(), commitToRollback)).findFirst.get() ?

n3nash commented on hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java:
```diff
@@ -169,7 +180,14 @@ public boolean archiveIfRequired(final JavaSparkContext jsc) throws IOException
+    // For archiving and cleaning instants, we need to include intermediate state files if they exist
+    HoodieActiveTimeline rawActiveTimeline = new HoodieActiveTimeline(metaClient, false);
+    Map<Pair<String, String>, List<HoodieInstant>> groupByTsAction = rawActiveTimeline.getInstants()
+        .collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
```
nit : s/x/i/g ?

n3nash commented on hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java:
```diff
@@ -169,7 +180,14 @@ public boolean archiveIfRequired(final JavaSparkContext jsc) throws IOException
+        .collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
+            x.getAction().equals(HoodieTimeline.COMPACTION_ACTION) ? HoodieTimeline.COMMIT_ACTION : x.getAction())));
+
+    return instants.flatMap(hoodieInstant ->
```
Should we filter only the necessary instants to group by  ?

rawActiveTimeline.getInstants().filter(i -> instants.contains(i)), this way the map will be smaller..

n3nash replied:
What I'm saying is group by only the instants chosen to archive, all types of actions is fine..

n3nash commented on hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java:
```diff
@@ -192,6 +210,7 @@ private boolean deleteArchivedInstants(List<HoodieInstant> archivedInstants) thr
       return i.isCompleted() && (i.getAction().equals(HoodieTimeline.COMMIT_ACTION)
           || (i.getAction().equals(HoodieTimeline.DELTA_COMMIT_ACTION)));
     }).max(Comparator.comparing(HoodieInstant::getTimestamp)));
+    log.info("Last Committed Instant =" + latestCommitted);
```
nit : space after "="

n3nash commented on hudi-client/src/test/java/org/apache/hudi/TestClientRollback.java:
```diff
@@ -217,10 +217,12 @@ public void testRollbackCommit() throws Exception {
 
       // simulate partial failure, where .inflight was not deleted, but data files were.
       HoodieTestUtils.createInflightCommitFiles(basePath, commitTime3);
+      Thread.sleep(1000);
```
Remove these now ?

n3nash commented on hudi-client/src/test/java/org/apache/hudi/index/TestHbaseIndex.java:
```diff
@@ -168,13 +169,19 @@ public void testTagLocationAndDuplicateUpdate() throws Exception {
     JavaRDD<HoodieRecord> javaRDD1 = index.tagLocation(writeRecords, jsc, hoodieTable);
+
     // Duplicate upsert and ensure correctness is maintained
+    // We are trying to approximately imitate the case when the RRD is recomputed. For RRD creating, driver code is not
```
Did not follow this, lets chat f2f

vinothchandar replied:
RDD :)

n3nash replied:
Discussed this f2f and resolved my doubts

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -420,17 +427,50 @@ public String getCommitActionType() {
+
+  public static Stream<HoodieInstant> dedupeInstants(Stream<HoodieInstant> instantStream) {
+    return instantStream.collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
+        x.getAction().equals(HoodieTimeline.COMPACTION_ACTION) ? HoodieTimeline.COMMIT_ACTION : x.getAction())))
```
Can we push this logic down to x.getAction ? For COMPACTION_ACTION, this should return COMMIT_ACTION, or implement a new method in HoodieInstant itself ?

bvaradar replied:
Refactored to reuse this logic.

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTimeline.java:
```diff
@@ -250,7 +253,10 @@ static HoodieInstant getCompactionInflightInstant(final String timestamp) {
   }
 
-  static HoodieInstant getInflightInstant(final HoodieInstant instant) {
+  static HoodieInstant getInflightInstant(final HoodieInstant instant, final HoodieTableType tableType) {
```
Is there a way to push this logic down to metaClient or HoodieInstant ?

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieInstant.java:
```diff
@@ -30,7 +34,23 @@
+  private static final Set<String> comparableActions = Arrays.stream(new String[] { HoodieTimeline.COMMIT_ACTION,
+      HoodieTimeline.COMPACTION_ACTION}).collect(Collectors.toSet());
+
+  public static final Comparator<HoodieInstant> comparator = Comparator.comparing(HoodieInstant::getTimestamp)
```
How does this ensure choosing the highest state ? Is it just string comparison ?

vinothchandar replied:
+1 on this

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -433,6 +494,32 @@ private void createFileInPath(Path fullPath, Option<byte[]> content) {
+   * @param fullPath File Path
+   * @param content Content to be stored
+   */
+  private void writeFileOnceInPath(Path fullPath, Option<byte[]> content) {
```
rename to `createImmutableFileInPath` ?

vinothchandar replied:
In other words why can't we use this method for the nullVersion too?

bvaradar replied:
@vinothchandar : Currently, we employ overwrite semantics for inflight commit/delta-commit files. We first create an empty inflight file and then overwrite it later with workload profile data. I can try cleaning that up in hoodie write client.  Let me know your thoughts ?

@n3nash : Renamed the method.

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -342,60 +365,94 @@ public HoodieInstant transitionCleanInflightToComplete(HoodieInstant inflightIns
+        }
+      } else {
+        // Ensures old state exists in timeline
+        System.out.println("Checking for file exists ?" + new Path(metaClient.getMetaPath(),
```
Convert to log

bvaradar replied:
Thanks. Fixed.

n3nash commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -342,60 +365,94 @@ public HoodieInstant transitionCleanInflightToComplete(HoodieInstant inflightIns
+        Path requestedInstantFilePath = new Path(metaClient.getMetaPath(),
+            new HoodieInstant(State.REQUESTED, inflight.getAction(), inflight.getTimestamp()).getFileName());
+
+        // If inflight and requested files do not exist, create one
```
what case will the inflight and requested not be present ?

bvaradar replied:
@n3nash : This can happen during the time of upgrade when the commit before we upgraded had to be reverted.

n3nash commented:
@bvaradar left some comments. In general, I couldn't understand how will existing tables move from VERSION_0 metadata to VERSION_1. Is the new version only supported for new tables ? If yes, what is the plan for the existing tables ?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTimeline.java:
```diff
@@ -250,7 +253,10 @@ static HoodieInstant getCompactionInflightInstant(final String timestamp) {
 
-  static HoodieInstant getInflightInstant(final HoodieInstant instant) {
+  static HoodieInstant getInflightInstant(final HoodieInstant instant, final HoodieTableType tableType) {
+    if ((tableType == HoodieTableType.MERGE_ON_READ) && instant.getAction().equals(COMMIT_ACTION)) {
```
add some comments for this special handling? I 'd prefer keeping this here compared to pushing this to HoodieInstant

bvaradar replied:
Added

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -151,17 +174,6 @@ public HoodieTimeline getDeltaCommitTimeline() {
         (Function<HoodieInstant, Option<byte[]>> & Serializable) this::getInstantDetails);
   }
 
-  /**
```
why the move? has code changed?

bvaradar replied:
Reverted.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -212,25 +224,24 @@ public void saveAsComplete(HoodieInstant instant, Option<byte[]> data) {
+  public HoodieInstant revertToInflight(HoodieInstant instant) {
+    log.info("Reverting instant to inflight " + instant);
+    HoodieInstant inflight = HoodieTimeline.getInflightInstant(instant, metaClient.getTableType());
+    revertCompleteToInflight(instant, inflight);
```
how do such reverts work now? we just seem to delete the complete instant and it will for e.g make it inflight again? is delete() atomic in cloud stores?

bvaradar replied:
Deletes are atomic in GCS but not so in S3. Thought about this. It shouldn't bring in new failure scenarios here.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -260,8 +271,13 @@ private void deleteInstantFile(HoodieInstant instant) {
 
-  public Option<byte[]> getInstantAuxiliaryDetails(HoodieInstant instant) {
-    Path detailPath = new Path(metaClient.getMetaAuxiliaryPath(), instant.getFileName());
+  public Option<byte[]> getPlanDetailsInBytes(HoodieInstant instant) {
```
rename: `readPlanAsBytes()` ?

bvaradar replied:
Done.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -260,8 +271,13 @@ private void deleteInstantFile(HoodieInstant instant) {
-  public Option<byte[]> getInstantAuxiliaryDetails(HoodieInstant instant) {
-    Path detailPath = new Path(metaClient.getMetaAuxiliaryPath(), instant.getFileName());
+  public Option<byte[]> getPlanDetailsInBytes(HoodieInstant instant) {
+    Path detailPath = null;
```
is null assignment superflous?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -312,8 +332,11 @@ public HoodieInstant transitionCompactionInflightToComplete(HoodieInstant inflig
-    Path fullPath = new Path(metaClient.getMetaAuxiliaryPath(), instant.getFileName());
-    createFileInPath(fullPath, data);
+    if (metaClient.getMetadataVersion().isNullVersion()) {
+      // No need for auxiliary folder for newer versions
```
add more context. it's not need only under a scenario right

bvaradar replied:
Done.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -404,12 +461,16 @@ public void saveToCleanRequested(HoodieInstant instant, Option<byte[]> content)
+    if (allowOverwrite || metaClient.getMetadataVersion().isNullVersion()) {
+      createFileInPath(fullPath, content);
+    } else {
+      writeFileOnceInPath(fullPath, content);
```
why do we have to support the two modes of creating? seems the difference is just that one creates and then opens, while one open a new file , writes, closes?

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieDefaultTimeline.java:
```diff
@@ -92,15 +91,15 @@ public HoodieTimeline filterInflightsAndRequested() {
+  public HoodieTimeline filterPendingExcludingCompaction() {
     return new HoodieDefaultTimeline(instants.stream().filter(instant -> {
-      return instant.isInflight() && (!instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION));
+      return (!instant.isCompleted()) && (!instant.getAction().equals(HoodieTimeline.COMPACTION_ACTION));
```
wont this provide both inflight and requested?

bvaradar replied:
Yes, that was the intention. One of the places where we are using is rolling back pending commits

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieInstant.java:
```diff
@@ -30,7 +34,23 @@
+   * A COMPACTION action eventually becomes COMMIT when completed. So, when grouping instants
+   * for state transitions, this needs to be taken into account
+   */
+  private static final Set<String> comparableActions = Arrays.stream(new String[] { HoodieTimeline.COMMIT_ACTION,
```
Hmm . seems most of the issue.stems from this switching of compaction => commit? Just throwing out an idea to see if we can just call talk about compaction as an implementation, but have the action be just commit? i.e remove Compaction action and replace with Commit given we have requested and inflight there? would that simplify the design? does it open new migration pains?

bvaradar replied:
Added a jira to track this : https://jira.apache.org/jira/browse/HUDI-384

vinothchandar commented on hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java:
```diff
@@ -169,7 +180,14 @@ public boolean archiveIfRequired(final JavaSparkContext jsc) throws IOException
+    HoodieActiveTimeline rawActiveTimeline = new HoodieActiveTimeline(metaClient, false);
+    Map<Pair<String, String>, List<HoodieInstant>> groupByTsAction = rawActiveTimeline.getInstants()
+        .collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
+            x.getAction().equals(HoodieTimeline.COMPACTION_ACTION) ? HoodieTimeline.COMMIT_ACTION : x.getAction())));
```
dont we have similar logic in timeline class itself? can we consolidate there?

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/io/HoodieCommitArchiveLog.java:
```diff
@@ -108,16 +110,25 @@ private void close() {
+      List<HoodieInstant> instantsToArchive = instantsToDelete.stream()
+          .filter(HoodieInstant::isCompleted).collect(Collectors.toList());
+
+      Preconditions.checkArgument(instantsToArchive.isEmpty() == instantsToDelete.isEmpty(),
```
Made code changes to allow pending instants also be archived.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/model/HoodieMetadataVersion.java:
```diff
@@ -0,0 +1,77 @@
+import java.util.Objects;
+
+/**
+ * Metadata Layout Version. Add new version when timeline format changes
```
Done

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java:
```diff
@@ -110,6 +114,10 @@ public static void createHoodieProperties(FileSystem fs, Path metadataFolder, Pr
       if (!properties.containsKey(HOODIE_ARCHIVELOG_FOLDER_PROP_NAME)) {
         properties.setProperty(HOODIE_ARCHIVELOG_FOLDER_PROP_NAME, DEFAULT_ARCHIVELOG_FOLDER);
       }
+      if (!properties.containsKey(HOODIE_TABLE_VERSION_PROP_NAME)) {
```
Added HoodieWriteConfig to control layout upgrade (default will be to upgrade to latest version). Redid the changes with the assumption that readers will be upgraded first.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -420,17 +427,50 @@ public String getCommitActionType() {
+   * @param fs                 FileSystem
+   * @param metaPath           Meta Path where hoodie instants are present
+   * @param includedExtensions Included hoodie extensions
+   * @param excludeIntermediateStates If there are multiple states for the same action instant,
```
Refactored this code to include TimelineLayout which handles the deduping part.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -420,17 +427,50 @@ public String getCommitActionType() {
+    return instantStream.sorted().collect(Collectors.toList());
+  }
+
+  public static Stream<HoodieInstant> dedupeInstants(Stream<HoodieInstant> instantStream) {
```
I moved this logic to TableLayout. Let me know if interface needs more changes.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -420,17 +427,50 @@ public String getCommitActionType() {
+  public static Stream<HoodieInstant> dedupeInstants(Stream<HoodieInstant> instantStream) {
+    return instantStream.collect(Collectors.groupingBy(x -> Pair.of(x.getTimestamp(),
+        x.getAction().equals(HoodieTimeline.COMPACTION_ACTION) ? HoodieTimeline.COMMIT_ACTION : x.getAction())))
+        .entrySet().stream().map(e -> e.getValue().stream().reduce((x, y) -> {
```
Fixed.

bvaradar commented:
@vinothchandar @n3nash : Redid the migration handling and addressed your comments

High-Level Changes since the previous review.
1. Hudi WriteConfig can be used to control timeline layout upgrade. Default will be to start avoiding renames in writer as soon as we deploy the new hudi jar on existing datasets. (unit-tests added)
2. Readers are assumed to upgrade before Writers
3. Added missing actions to archiving. Now, Archiving logs all states of each action
4. Changed some code in Cleaner Migration (cc @leesf  - please take a look).  If this is too painful to review, I will try to create a separate diff.

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableConfig.java:
```diff
@@ -112,6 +116,10 @@ public static void createHoodieProperties(FileSystem fs, Path metadataFolder, Pr
       if (!properties.containsKey(HOODIE_ARCHIVELOG_FOLDER_PROP_NAME)) {
         properties.setProperty(HOODIE_ARCHIVELOG_FOLDER_PROP_NAME, DEFAULT_ARCHIVELOG_FOLDER);
       }
+      if (!properties.containsKey(HOODIE_TIMELINE_LAYOUT_VERSION)) {
```
Does this mean that only new datasets will get version 1? What happens to existing datasets? They stick to version 0?

bvaradar replied:
The version in hoodie.properties is treated as default configuration. MetaClient Writers override this config with version-1 (unless they opt out). MetaClient readers will always use latest version (version - 1) which is backwards compatible. Keeping a version in hoodie.properties would also hopefully be useful in future migrations

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/model/TimelineLayoutVersion.java:
```diff
@@ -0,0 +1,79 @@
+  public static final Integer VERSION_1 = 1; // current version with no renames
+
+  public static final Integer CURR_VERSION = VERSION_1;
+  public static final TimelineLayoutVersion LATEST_TIMELINE_LAYOUT_VERSION = new TimelineLayoutVersion(CURR_VERSION);
```
rename to `CURR_LAYOUT_VERSION`

bvaradar replied:
Done

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -274,25 +286,35 @@ public synchronized HoodieArchivedTimeline getArchivedTimeline() {
     properties.setProperty(HoodieTableConfig.HOODIE_TABLE_NAME_PROP_NAME, tableName);
     properties.setProperty(HoodieTableConfig.HOODIE_TABLE_TYPE_PROP_NAME, tableType.name());
-    if (tableType == HoodieTableType.MERGE_ON_READ) {
+    if ((tableType == HoodieTableType.MERGE_ON_READ) && (payloadClassName != null)) {
```
Redundant braces?

bvaradar replied:
Done

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -87,11 +90,14 @@ public HoodieTableMetaClient(Configuration conf, String basePath) throws Dataset
   public HoodieTableMetaClient(Configuration conf, String basePath, boolean loadActiveTimelineOnLoad) {
-    this(conf, basePath, loadActiveTimelineOnLoad, ConsistencyGuardConfig.newBuilder().build());
+    this(conf, basePath, loadActiveTimelineOnLoad, ConsistencyGuardConfig.newBuilder().build(),
+        // Readers will use latest version
```
Even writers use `HoodieTableMetaClient` right? can you clarify this comment?

bvaradar replied:
Agree, this is misleading. What I meant was MetaClient Readers ( use-cases which just lists the .hoodie folder) as opposed to MetaClient Writer (performing action transitions in .hoodie folder). Will remove this comment

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/HoodieTableMetaClient.java:
```diff
@@ -414,23 +436,41 @@ public String getCommitActionType() {
+            })).map(HoodieInstant::new);
+
+    if (applyLayoutVersionFilters) {
+      instantStream = TimelineLayout.getLayout(getTimelineLayoutVersion()).filterHoodieInstants(instantStream);
```
Seems the `applyLayoutVersionFilters` is set selectively using which HoodieActiveTimeline constructor is invoked? Would this be fragile.. Thinking out loud, applying filters on V0, has no effect since there are nothing to get rid off. Only thing that could do wrong is not filtering V1.. hmmm

bvaradar replied:
The key case here is for archival where we need all instants without filtering. May be introduce couple of factory methods which instantiate HoodieActiveTimeline w/o filtering ?  HUDI-414

vinothchandar commented on hudi-common/src/main/java/org/apache/hudi/common/table/TimelineLayout.java:
```diff
@@ -0,0 +1,79 @@
+  /**
+   * Table Layout where state transitions are managed by renaming files.
+   */
+  private static class TimelineLayoutV0 extends TimelineLayout {
```
lets add a small unit tests for this, if it doesnot exist already

bvaradar replied:
Done

bvaradar commented on hudi-cli/src/main/java/org/apache/hudi/cli/commands/DatasetsCommand.java:
```diff
@@ -85,6 +86,8 @@ public String createTable(
       @CliOption(key = {"tableType"}, unspecifiedDefaultValue = "COPY_ON_WRITE",
           help = "Hoodie Table Type. Must be one of : COPY_ON_WRITE or MERGE_ON_READ") final String tableTypeStr,
+      @CliOption(key = {"archiveLogFolder"}, help = "Folder Name for storing archived timeline") String archiveFolder,
+      @CliOption(key = {"layoutVersion"}, help = "Specific Layouyt Version to use") Integer layoutVersion,
```
Done

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/HoodieWriteClient.java:
```diff
@@ -792,23 +798,25 @@ public void restoreToInstant(final String instantTime) throws HoodieRollbackExce
         .collect(Collectors.toList());
     // Start a rollback instant for all commits to be rolled back
-    String startRollbackInstant = startInstant();
+    String startRollbackInstant = HoodieActiveTimeline.createNewCommitTime();
```
Done

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieInstant.java:
```diff
@@ -32,7 +35,24 @@
+  public static final Comparator<HoodieInstant> COMPARATOR = Comparator.comparing(HoodieInstant::getTimestamp)
+      .thenComparing(ACTION_COMPARATOR).thenComparing(HoodieInstant::getState);
+
+  public static final String getCompatibleAction(String action) {
```
Picked Comparable.

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -61,11 +62,11 @@
-          REQUESTED_COMPACTION_EXTENSION, INFLIGHT_RESTORE_EXTENSION, RESTORE_EXTENSION}));
+  public static final Set<String> VALID_EXTENSIONS_IN_ACTIVE_TIMELINE = new HashSet<>(Arrays.asList(
+      new String[]{COMMIT_EXTENSION, INFLIGHT_COMMIT_EXTENSION, REQUESTED_COMMIT_EXTENTSION, DELTA_COMMIT_EXTENSION,
+          INFLIGHT_DELTA_COMMIT_EXTENSION, REQUESTED_DELTA_COMMIT_EXTENTSION, SAVEPOINT_EXTENSION,
```
Done

bvaradar commented on hudi-common/src/main/java/org/apache/hudi/common/table/timeline/HoodieActiveTimeline.java:
```diff
@@ -61,11 +62,11 @@
-          INFLIGHT_CLEAN_EXTENSION, REQUESTED_CLEAN_EXTENSION, INFLIGHT_COMPACTION_EXTENSION,
-          REQUESTED_COMPACTION_EXTENSION, INFLIGHT_RESTORE_EXTENSION, RESTORE_EXTENSION}));
+  public static final Set<String> VALID_EXTENSIONS_IN_ACTIVE_TIMELINE = new HashSet<>(Arrays.asList(
+      new String[]{COMMIT_EXTENSION, INFLIGHT_COMMIT_EXTENSION, REQUESTED_COMMIT_EXTENTSION, DELTA_COMMIT_EXTENSION,
```
Done

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -48,6 +49,7 @@
 public class HoodieWriteConfig extends DefaultHoodieConfig {
 
   public static final String TABLE_NAME = "hoodie.table.name";
+  private static final String OVERRIDDEN_TIMELINE_LAYOUT_VERSION = "hoodie.timeline.layout.version";
```
Done

bvaradar commented on hudi-client/src/main/java/org/apache/hudi/config/HoodieWriteConfig.java:
```diff
@@ -733,6 +746,11 @@ public HoodieWriteConfig build() {
       setDefaultOnCondition(props, !isConsistencyGuardSet,
           ConsistencyGuardConfig.newBuilder().fromProperties(props).build());
 
+      String layoutVersion = props.getProperty(OVERRIDDEN_TIMELINE_LAYOUT_VERSION);
```
Done.

bvaradar commented on hudi-common/src/main/avro/HoodieArchivedMetaEntry.avsc:
```diff
@@ -74,6 +75,27 @@
+         "default": null
+      },
+      {
+         "name":"actionState",
```
Yes, both actionType and instantTime are already available.

bvaradar commented on hudi-common/src/main/avro/HoodieArchivedMetaEntry.avsc:
```diff
@@ -37,6 +37,7 @@
          "default": null
       },
       {
+         /** DEPRECATED **/
```
Yeah, This is was intended to be equivalent of HoodieCommitMetadata but since we are using HoodieCommitMetadata for storing results of compaction, this is never used.

vinothchandar commented:
@bvaradar Feel free to merge when you feel this is ready
